samza-sql-shell/src/main/java/org/apache/samza/sql/client/cli/CliEnvironment.java

```diff
@@ -19,112 +19,168 @@
+        executorEnvHandler = executor.getEnvironmentVariableHandler();
+        return 0;
+      } else {
+        return -2;
+      }
     }
 
-    return 0;
+    EnvironmentVariableHandler handler = getAppropriateHandler(envName);
+    if(handler == null) {
+      // Shell doesn't recognize this variable. There's no executor handler yet. Save for future executor
+      if(delayedExecutorVars == null) {
+        delayedExecutorVars = new HashMap<>();
+      }
+      delayedExecutorVars.put(envName, value);
+      return 0;
+    }
+
+    return handler.setEnvironmentVariable(envName, value);
   }
 
-  // TODO: Separate the values out of the logic part
-  List<String> getPossibleValues(String var) {
-    List<String> vals = new ArrayList<>();
-    switch (var.toLowerCase()) {
-      case debugEnvVar:
-        vals.add("true");
-        vals.add("false");
-        return vals;
-      default:
-        return null;
+  public String getEnvironmentVariable(String envName) {
+    if(envName.equalsIgnoreCase(CliConstants.CONFIG_EXECUTOR)) {
+      return activeExecutorClassName;
     }
-  }
 
-  void printAll(Writer writer) throws IOException {
-    writer.write(debugEnvVar);
-    writer.write('=');
-    writer.write(debug.toString());
-    writer.write('\n');
-  }
+    EnvironmentVariableHandler handler = getAppropriateHandler(envName);
+    if(handler == null)
+      return null;
 
-  private void disableJavaSystemOutAndErr() {
-    PrintStream ps = new PrintStream(new NullOutputStream());
-    System.setOut(ps);
-    System.setErr(ps);
+    return handler.getEnvironmentVariable(envName);
   }
 
-  private void enableJavaSystemOutAndErr() {
-    System.setOut(stdout);
-    System.setErr(stderr);
+  public String[] getPossibleValues(String envName)
+  {
+    EnvironmentVariableHandler handler = getAppropriateHandler(envName);
+    if(handler == null)
+      return null;
+
+    EnvironmentVariableSpecs.Spec spec = handler.getSpecs().getSpec(envName);
+    return spec == null ? null : spec.getPossibleValues();
   }
 
-  void takeEffect() {
-    if (debug) {
-      enableJavaSystemOutAndErr();
-    } else {
-      // We control terminal directly; Forbid any Java System.out and System.err stuff so
-      // any underlying output will not mess up the console
-      disableJavaSystemOutAndErr();
+  public void printAll(PrintWriter writer) {
+    printVariable(writer, CliConstants.CONFIG_EXECUTOR, activeExecutorClassName);
+    printAll(writer, shellEnvHandler);
+    if(executorEnvHandler != null) {
+      printAll(writer, executorEnvHandler);
     }
   }
 
-  private class NullOutputStream extends OutputStream {
-    public void close() {
+  /**
+   * Gives CliEnvironment a chance to apply settings, especially during initialization, things like
+   * making default values take effect
+   */
+  public void finishInitialization() {
+    if(executor == null) {
+      if(createShellExecutor(CliConstants.DEFAULT_EXECUTOR_CLASS)) {
```

Comment:
rather than have the method return boolean do you think it is a good idea to have the exception thrown and logged below?

Reply:
Good point @weiqingy  As we discussed offline, Let's do regular exceptions for all the internal methods and in the outermost layer before the information is returned to the user, You can handle these exceptions and transform them into user understandable form.